### PR TITLE
feat: menu support css variable theme

### PR DESCRIPTION
**Pull request by kiner-tang**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [X] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- #45618

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | menu support css variable theme |
| 🇨🇳 Chinese | 菜单支持 css 变量主题 |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [X] Doc is updated/provided or not needed
- [X] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 72681e7</samp>

This pull request adds CSS variable support to the menu component by using a custom hook and a utility library. It refactors the menu component style tokens and CSS objects to use the utility functions for handling units, arithmetic operations, and type aliases. It also fixes a typo in the RTL mode and improves the compatibility with different CSS units. The main files affected are `components/menu/menu.tsx`, `components/menu/style/index.tsx`, and `components/menu/style/cssVar.ts`.

### <samp>🤖 Generated by Copilot at 72681e7</samp>

*  Add a custom hook `useCSSVar` to register and use CSS variables for the menu component ([link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-4451508a2fd1c741b90ee521fda79d7008d64294f2e5f23c46c678f76c1e32cdR1-R8))
*  Import and call `useCSSVar` in `menu.tsx`, `MenuDivider.tsx`, `MenuItem.tsx`, and `SubMenu.tsx` and wrap the returned elements with the CSS variable provider ([link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-bcc9a016f0b5e2844ec9d0aa399d06a86134f535931eeced33c3d08c3462ce43R21),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-bcc9a016f0b5e2844ec9d0aa399d06a86134f535931eeced33c3d08c3462ce43L123-R125),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-bcc9a016f0b5e2844ec9d0aa399d06a86134f535931eeced33c3d08c3462ce43L159-R161),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-363da270043c53f8d549bddff27babd402b8262441c97573422d320a7003f35dR5),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-363da270043c53f8d549bddff27babd402b8262441c97573422d320a7003f35dR19),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-363da270043c53f8d549bddff27babd402b8262441c97573422d320a7003f35dL25-R27),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-871ad818189b22c5d499c18f2e2d8905b9b6ff5e8fb01bc8008d57ffe2aa8509R14),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-871ad818189b22c5d499c18f2e2d8905b9b6ff5e8fb01bc8008d57ffe2aa8509R60-R61),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-871ad818189b22c5d499c18f2e2d8905b9b6ff5e8fb01bc8008d57ffe2aa8509L113-R116),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-78ccc41cf5d14904062e19511837d0f4b6c9926376efc52facefc386c21bac84R10),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-78ccc41cf5d14904062e19511837d0f4b6c9926376efc52facefc386c21bac84R38),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-78ccc41cf5d14904062e19511837d0f4b6c9926376efc52facefc386c21bac84L73-R75),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-78ccc41cf5d14904062e19511837d0f4b6c9926376efc52facefc386c21bac84L87-R89))
*  Import and use `unit` and `type` functions from `@ant-design/cssinjs` to add unit suffixes to numeric values and declare type aliases for CSS objects in `index.tsx`, `horizontal.tsx`, `vertical.tsx`, and `theme.tsx` ([link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL1-R11),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-ebae966ac1b7399f66aa822d0ead52d0e27fcc1d5d23dfc05eaf39125e693467R1),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-803a82d9b4afa684620a38d6c44836f0b24b12205ef42ec80dba78aa28fe9829L1-R1),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-b7b7775fd3e73c4510f6e127d709629338c94f257ad4cde92ffa63c62f11e25bL1-R1))
*  Modify the token types and values in `index.tsx` to allow unitless values and use `token.calc` and `unit` functions to perform arithmetic operations and add unit suffixes ([link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL38-R43),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL363-R369),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL464-R469),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL474-R480),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL488-R495),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL573-R580),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL649-R656),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL670-R677),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL767-R776),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL774-R792),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL830-R966),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL897-R1027))
*  Add a new function `prepareComponentToken` in `index.tsx` to return the default token values for the menu component with dark mode and disabled state colors ([link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dR809-R927))
*  Replace the original object literal with the default token values for the menu component with the `prepareComponentToken` function call in `index.tsx` ([link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL830-R966),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f84706d0f4fb54c91e83f8dd7b34ce37f6015d56dc057385580b19473c8f1f7dL897-R1027))
*  Fix a typo in the `transform` value for the submenu arrow in RTL mode in `rtl.tsx` ([link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-f85ad6b602813fba88abb12685543822dacf9af29552d875a0f46b353cdc9ad6L18-R18))
*  Replace the original token values with `token.calc` and `unit` functions in `theme.tsx` to perform arithmetic operations and add unit suffixes ([link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-803a82d9b4afa684620a38d6c44836f0b24b12205ef42ec80dba78aa28fe9829L175-R175),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-803a82d9b4afa684620a38d6c44836f0b24b12205ef42ec80dba78aa28fe9829L183-R183),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-803a82d9b4afa684620a38d6c44836f0b24b12205ef42ec80dba78aa28fe9829L213-R213),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-803a82d9b4afa684620a38d6c44836f0b24b12205ef42ec80dba78aa28fe9829L228-R228),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-803a82d9b4afa684620a38d6c44836f0b24b12205ef42ec80dba78aa28fe9829L239-R239))
*  Replace the original token values with `token.calc` and `unit` functions in `horizontal.tsx` and `vertical.tsx` to perform arithmetic operations and add unit suffixes ([link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-ebae966ac1b7399f66aa822d0ead52d0e27fcc1d5d23dfc05eaf39125e693467L19-R20),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-b7b7775fd3e73c4510f6e127d709629338c94f257ad4cde92ffa63c62f11e25bL17-R17),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-b7b7775fd3e73c4510f6e127d709629338c94f257ad4cde92ffa63c62f11e25bL27-R27),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-b7b7775fd3e73c4510f6e127d709629338c94f257ad4cde92ffa63c62f11e25bL33-R33),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-b7b7775fd3e73c4510f6e127d709629338c94f257ad4cde92ffa63c62f11e25bL39-R39),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-b7b7775fd3e73c4510f6e127d709629338c94f257ad4cde92ffa63c62f11e25bL71-R71),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-b7b7775fd3e73c4510f6e127d709629338c94f257ad4cde92ffa63c62f11e25bL100-R100),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-b7b7775fd3e73c4510f6e127d709629338c94f257ad4cde92ffa63c62f11e25bL181-R183),[link](https://github.com/ant-design/ant-design/pull/45750/files?diff=unified&w=0#diff-b7b7775fd3e73c4510f6e127d709629338c94f257ad4cde92ffa63c62f11e25bL194-R196))

### CODE REVIEW

**Comment by MadCcc** [comment]
argos 上有变化了

File: components/menu/MenuDivider.tsx [hunk 1]
```diff
@@ -15,14 +16,15 @@ const MenuDivider: React.FC<MenuDividerProps> = (props) => {
   const { getPrefixCls } = React.useContext(ConfigContext);
 
   const prefixCls = getPrefixCls('menu', customizePrefixCls);
+  const wrapCSSVar = useCSSVar(prefixCls);
```
Comment (MadCcc) [inline components/menu/MenuDivider.tsx:19]: 这里可以不用写，不是每一个子组件都需要的，和 useStyle 同时出现就可以了

File: components/menu/MenuItem.tsx [hunk 2]
```diff
@@ -56,6 +57,8 @@ const MenuItem: GenericComponent = (props) => {
     return wrapNode;
   };
 
+  const wrapCSSVar = useCSSVar(prefixCls);
```
Comment (MadCcc) [inline components/menu/MenuItem.tsx:60]: 同上

File: components/menu/style/index.tsx [hunk 3]
```diff
@@ -471,8 +476,8 @@ const genSubMenuArrowStyle = (token: MenuToken): CSSObject => {
         // →
         '&::before, &::after': {
           position: 'absolute',
-          width: menuArrowSize * 0.6,
-          height: menuArrowSize * 0.15,
+          width: unit(token.calc(menuArrowSize).mul(0.6).equal()),
+          height: unit(token.calc(menuArrowSize).mul(0.15).equal()),
```
Comment (MadCcc) [inline components/menu/style/index.tsx:480]: 不用加 unit

File: components/menu/style/index.tsx [hunk 4]
```diff
@@ -461,7 +466,7 @@ const genSubMenuArrowStyle = (token: MenuToken): CSSObject => {
         position: 'absolute',
         top: '50%',
         insetInlineEnd: token.margin,
-        width: menuArrowSize,
+        width: unit(menuArrowSize),
```
Comment (MadCcc) [inline components/menu/style/index.tsx:469]: 不用加 unit

File: components/menu/style/index.tsx [hunk 5]
```diff
@@ -827,13 +957,13 @@ export default (prefixCls: string, injectStyle: boolean): UseComponentStyleResul
...
-      const menuArrowSize = (fontSize / 7) * 5;
+      const menuArrowSize = token.calc(fontSize).div(7).mul(5).equal();
 
       // Menu Token
       const menuToken = mergeToken<MenuToken>(token, {
         menuArrowSize,
-        menuHorizontalHeight: controlHeightLG * 1.15,
-        menuArrowOffset: `${menuArrowSize * 0.25}px`,
+        menuHorizontalHeight: token.calc(controlHeightLG).mul(1.15).equal(),
+        menuArrowOffset: unit(token.calc(menuArrowSize).mul(0.25).equal()),
```
Comment (MadCcc) [inline components/menu/style/index.tsx:966]: 这里不要加 unit 了，上面使用的地方都用了 unit，这里就当数值就可以

File: components/menu/style/rtl.tsx [hunk 6]
```diff
@@ -15,7 +15,7 @@ const getRTLStyle: GenerateStyle<MenuToken> = ({ componentCls, menuArrowOffset }
     ${componentCls}-submenu-rtl ${componentCls}-vertical`]: {
     [`${componentCls}-submenu-arrow`]: {
       '&::before': {
-        transform: `rotate(-45deg) translateY(-${menuArrowOffset})`,
+        transform: `rotate(-45deg) translateY(${menuArrowOffset})`,
```
Comment (MadCcc) [inline components/menu/style/rtl.tsx:18]: 注意下 menuArrowOffset 是否带单位

File: components/menu/SubMenu.tsx [hunk 7]
```diff
@@ -70,7 +72,7 @@ const SubMenu: React.FC<SubMenuProps> = (props) => {
   // ============================ zIndex ============================
   const [zIndex] = useZIndex('Menu');
 
-  return (
+  return wrapCSSVar(
```
Comment (MadCcc) [inline components/menu/SubMenu.tsx:75]: SubMenu 应该不用包，这个组件不会单独使用

Comment (kiner-tang) [reply components/menu/SubMenu.tsx:75]: 好像需要的，不然样式会丢

Comment (kiner-tang) [reply components/menu/SubMenu.tsx]: 好像去掉确实可以，之前应该是其他原因影响了

**Review by MadCcc — APPROVED** [review]

**Pull request merged**